### DISCUSSION TITLE
Test failing for testUserCanFundInteractionsFundMe

### DISCUSSION THREAD

**Discussion by shrutic11** [Cyfrin/foundry-full-course-cu#397] [Q&A] [answered]
I decided to go ahead an write two separate test cases to test the interactions for FundFundMe and WithdrawFundMe.

My WithdrawFundMe test is passing, however, the FundFundMe test case is failing with the following error:

```
Running 1 test for test/integration/InteractionsTest.t.sol:InteractionsTest
[FAIL. Reason: Assertion failed.] testUserCanFundInteractionsFundMe() (gas: 385275)
Logs:
  Error: a == b not satisfied [address]
        Left: 0x1804c8ab1f12e6bbf3894d4083f33e07309d1f38
       Right: 0xbb95e6c56f65ab04ee1d47646b3326fa07a03fe3

Traces:
  [2933702] InteractionsTest::setUp() 
    ├─ [1463142] → new DeployFundMe@0xFEfC6BAF87cF3684058D62Da40Ff3A795946Ab06
    │   └─ ← 7087 bytes of code
    ├─ [1411812] DeployFundMe::run() 
    │   ├─ [916827] → new HelperConfig@0x6d2eed85750d316088343D6d5e91ca59eb052768
    │   │   ├─ [0] VM::startBroadcast() 
    │   │   │   └─ ← ()
    │   │   ├─ [372255] → new MockV3Aggregator@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   └─ ← 1082 bytes of code
    │   │   ├─ [0] VM::stopBroadcast() 
    │   │   │   └─ ← ()
    │   │   └─ ← 2209 bytes of code
    │   ├─ [359] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← MockV3Aggregator: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]
    │   ├─ [0] VM::startBroadcast() 
    │   │   └─ ← ()
    │   ├─ [428033] → new FundMe@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   └─ ← 2026 bytes of code
    │   ├─ [0] VM::stopBroadcast() 
    │   │   └─ ← ()
    │   └─ ← FundMe: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]
    ├─ [0] VM::deal(TEST_USER: [0xBb95e6c56f65AB04ee1D47646b3326Fa07a03fe3], 10000000000000000000 [1e19]) 
    │   └─ ← ()
    └─ ← ()

  [385275] InteractionsTest::testUserCanFundInteractionsFundMe() 
    ├─ [0] VM::prank(TEST_USER: [0xBb95e6c56f65AB04ee1D47646b3326Fa07a03fe3]) 
    │   └─ ← ()
    ├─ [239901] → new FundFundMe@0x563b9bE99c6291C039DF72bFf4ED5bb2dDfc5003
    │   └─ ← 977 bytes of code
    ├─ [92010] FundFundMe::fundFundMe(FundMe: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]) 
    │   ├─ [0] VM::startBroadcast() 
    │   │   └─ ← ()
    │   ├─ [81388] FundMe::sendUSD{value: 10000000000000000}() 
    │   │   ├─ [8993] MockV3Aggregator::latestRoundData() [staticcall]
    │   │   │   └─ ← 1, 200000000000, 1, 1, 1
    │   │   └─ ← ()
    │   ├─ [0] VM::stopBroadcast() 
    │   │   └─ ← ()
    │   └─ ← ()
    ├─ [642] FundMe::getFunders(0) [staticcall]
    │   └─ ← DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38]
    ├─ emit log(: Error: a == b not satisfied [address])
    ├─ emit log_named_address(key:       Left, val: DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
    ├─ emit log_named_address(key:      Right, val: TEST_USER: [0xBb95e6c56f65AB04ee1D47646b3326Fa07a03fe3])
    ├─ [0] VM::store(VM: [0x7109709ECfa91a80626fF3989D68f67F5b1DD12D], 0x6661696c65640000000000000000000000000000000000000000000000000000, 0x0000000000000000000000000000000000000000000000000000000000000001) 
    │   └─ ← ()
    └─ ← ()

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 1.41ms
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)
```

Below is my InteractionsTest.t.sol

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../../src/FundMe.sol";
import {DeployFundMe} from "../../script/DeployFundMe.s.sol";
import {FundFundMe, WithdrawFundMe} from "../../script/Interactions.s.sol";

contract InteractionsTest is Test {
    FundMe fundMe;
    address TEST_USER = makeAddr("TEST_USER");
    uint256 constant TEST_SEND_ETH = 0.01 ether;
    uint256 constant STARTING_BALANCE = 10 ether;

    function setUp() external {
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
        vm.deal(TEST_USER, STARTING_BALANCE);
    }

    function testUserCanFundInteractionsFundMe() public {
        vm.prank(TEST_USER);
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe));

        address funder = fundMe.getFunders(0);
        assertEq(funder, TEST_USER);
    }

    function testUserCanFundInteractionsWithdraw() public {
        vm.prank(TEST_USER);
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe));

        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
        withdrawFundMe.withdrawFundMe(address(fundMe));

        assert(address(fundMe).balance == 0);
    }
}
```

Here's my FundFundMeContract
```Solidity
contract FundFundMe is Script {
    uint256 public constant SEND_ETH_VALUE = 0.01 ether;

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );
        vm.startBroadcast();
        fundFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }

    function fundFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).sendUSD{value: SEND_ETH_VALUE}();

        vm.stopBroadcast();
    }
}
```
I want to understand, if my FundFundMe is successful with the TEST USER created using vm.prank("TEST_USER"), why is the assertion failing?

**Comment by shikhar229169** [comment, marked as answer]
In the fundMe interaction script we are using vm.startBroadcast(), which means that the transactions will be performed on behalf of that account for which you passed the private key by --private-key.
If you run ```forge test```, it will temporarily run a local chain and passes a private key on its own. So the funding was actually performed on behalf of that particular account.
So, as a result of which its showing its different.

For this to work, instead of comparing with TEST_USER, you should compare with ```msg.sender```.

One more thing ```vm.prank(TEST_USER)``` only works for the immediate next line.
Also remember that we can't use broadcast and prank simultaneously.

**Reply by shrutic11** [reply]
Yes, asserting the funder with msg.sender is indeed working. But I want to understand how and what changes I need to make in the tests to execute it with a test user. 

If I remove vm.broadcast from the FundFundMe contract:: fundFundMe function, I get EVM : OutOfFund error, but in this case the user is indeed the test_user created by vm.prank. And even though I can see it has all the funds, I don't understand why I am getting out of fund error.

Running 1 test for test/integration/InteractionsTest.t.sol:InteractionsTest
[FAIL. Reason: EvmError: Revert] testUserCanFundInteractionsFundMe() (gas: 255831)
Traces:
  [255831] InteractionsTest::testUserCanFundInteractionsFundMe() 
    ├─ [206071] → new FundFundMe@0x2a9e8fa175F45b235efDdD97d2727741EF4Eee63
    │   └─ ← 808 bytes of code
    ├─ [0] VM::prank(TEST_USER: [0xBb95e6c56f65AB04ee1D47646b3326Fa07a03fe3]) 
    │   └─ ← ()
    ├─ [9848] FundFundMe::fundFundMe(FundMe: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]) 
    │   ├─ [0] FundMe::sendUSD{value: 10000000000000000}() 
    │   │   └─ ← "EvmError: OutOfFund"
    │   └─ ← "EvmError: Revert"
    └─ ← "EvmError: Revert"

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 881.08µs
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

modified Interactions.s.sol
```solidity
contract FundFundMe is Script {
    uint256 public constant SEND_ETH_VALUE = 0.01 ether;

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );
        vm.startBroadcast();
        fundFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }

    function fundFundMe(address mostRecentlyDeployed) public {
        // vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).sendUSD{value: SEND_ETH_VALUE}();
        //  vm.stopBroadcast();
    }
}
```

modified InteractionsTest.t.sol;
```solidity
   function setUp() external {
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
        vm.deal(TEST_USER, STARTING_BALANCE);
    }

    function testUserCanFundInteractionsFundMe() public {
        FundFundMe fundFundMe = new FundFundMe();
        vm.prank(TEST_USER);
        fundFundMe.fundFundMe(address(fundMe));

        address funder = fundMe.getFunders(0);
        assertEq(funder, TEST_USER);
    }
```

**Reply by shikhar229169** [reply]
When you remove start broadcast, then the one who is calling the ```sendUSD``` function is the ```FundFundMe``` contract, as the contract has 0 ETH, thus its reverting with out of fund error.
The below funtion was no doubt called by ```TEST_USER``` but the one who calls for ```sendUSD``` function was the ```FundFundMe``` contract.
```solidity
fundFundMe.fundFundMe(address(fundMe));
``` 

I recommend you to perform some console.log with these address stuffs and for sure it will clear your concepts.

**Reply by shikhar229169** [reply]
So, if you want it to work with test user, then you need to make change in your interactions, by adding prank for TEST_USER
```solidity
function fundFundMe(address mostRecentlyDeployed) public {
        vm.prank(TEST_USER)
        FundMe(payable(mostRecentlyDeployed)).sendUSD{value: SEND_ETH_VALUE}();
}
```
But according to me this is not a good practice and should never be done in Interactions script.
In interactions we actually want the one with the --private-key arg should perform the transactions where ever we add broadcast, so for tests we should check for the same user, which will be ```msg.sender``` in the tests.

**Reply by shrutic11** [reply]
I agree. No, I don't want to add vm.prank in the Interactions script either. I will work through it ans try to understand. I will post the answer here once I have a solution. Thanks for the guidance.

**Reply by Imad-Allal** [reply]
So did you get any solution ?
